kpi/views/v2/data.py

```diff
@@ -549,3 +546,89 @@ def _filter_mongo_query(self, request):
             )
 
         return filters
+
+    def _get_enketo_link(
+        self, request: Request, submission_id: int, action_: str
+    ) -> Response:
+
+        deployment = self._get_deployment()
+        user = request.user
+
+        if action_ == 'edit':
+            partial_perm = PERM_CHANGE_SUBMISSIONS
+        elif action_ == 'view':
+            partial_perm = PERM_VIEW_SUBMISSIONS
+
+        # User's permissions are validated by the permission class. This extra step
+        # is needed to validate at a row level for users with partial permissions.
+        # A `PermissionDenied` error will be raised if it is not the case.
+        # `validate_access_with_partial_perms()` is called no matter what are the
+        # user's permissions. The first check inside the method is the user's
+        # permissions. `submission_ids` should be equal to `None` if user has
+        # regular permissions.
+        deployment.validate_access_with_partial_perms(
+            user=user,
+            perm=partial_perm,
+            submission_ids=[submission_id],
+        )
+
+        # The XML version is needed for Enketo
+        submission_xml = deployment.get_submission(
+            submission_id, user, SUBMISSION_FORMAT_TYPE_XML
+        )
+        # The JSON version is needed to detect its version
+        submission_json = deployment.get_submission(
+            submission_id, user, request=request
+        )
+
+        _, submissions_stream = build_formpack(
+            self.asset,
+            submission_stream=[submission_json],
+            use_all_form_versions=True
+        )
+        version_uid = list(submissions_stream)[0][INFERRED_VERSION_ID_KEY]
+
+        try:
+            snapshot = self.asset.versioned_snapshot(version_uid=version_uid)
+        except AssetVersion.DoesNotExist:
+            raise serializers.ValidationError(
+                {'version': t('Version not found')}
+            )
```

Comment:
i don't think this should ever happen since `_infer_version_id()` falls back on the latest deployed version. i suppose there could be a race condition between `_infer_version_id()` and `versioned_snapshot()`, but even then, perhaps we should return a 500 error. it's not something the client can correct.